vmmlib/frustum.hpp

```diff
@@ -135,177 +139,109 @@ typedef frustum< double > Frustumd;
 namespace vmml
 {
 
-template< typename T >
-const frustum< T > frustum< T >::DEFAULT( static_cast< T >( -1.0 ),
-                                          static_cast< T >( 1.0 ),
-                                          static_cast< T >( -1.0 ),
-                                          static_cast< T >( 1.0 ),
-                                          static_cast< T >( 0.1 ),
-                                          static_cast< T >( 100.0 ) );
-
-
-template < typename T >
-frustum<T>::frustum( const T _left, const T _right, const T _bottom,
-                     const T _top, const T _near, const T _far )
+template< typename T > Frustum< T >::Frustum()
 {
-    set( _left, _right, _bottom, _top, _near, _far );
+    _array[0] = -1;
+    _array[1] = 1;
+    _array[2] = -1;
+    _array[3] = 1;
+    _array[4] = 0.1;
+    _array[5] = 100;
 }
 
-
 template < typename T >
-template< typename U >
-frustum< T >::frustum( const frustum< U >& source_ )
+Frustum<T>::Frustum( const T _left, const T _right, const T _bottom,
+                     const T _top, const T _near, const T _far )
 {
-    (*this) = source_;
+    _array[0] = _left;
+    _array[1] = _right;
+    _array[2] = _bottom;
+    _array[3] = _top;
+    _array[4] = _near;
+    _array[5] = _far;
 }
 
-
-
 template < typename T >
-template< typename U >
-frustum< T >::frustum( const U* values )
+Frustum<T>::Frustum( const T fov_y, const T aspect_ratio, const T nearPlane_,
+                     const T farPlane_ )
 {
-    assert( values &&
-            "frustum: Nullpointer argument as source for initialisation!" );
-    left()      = static_cast< T > ( values[0] );
-    right()     = static_cast< T > ( values[1] );
-    bottom()    = static_cast< T > ( values[2] );
-    top()       = static_cast< T > ( values[3] );
-    near_plane() = static_cast< T > ( values[4] );
-    far_plane()  = static_cast< T > ( values[5] );
+    _array[2] = std::tan( 0.5 * fov_y * M_PI / 180.0 ) * 0.5;
+    _array[3] = -_array[2];
+    // depend on _array[2,3]:
+    _array[0] = bottom() * aspect_ratio;
+    _array[1] = top() * aspect_ratio;
+    _array[4] = nearPlane_;
+    _array[5] = farPlane_;
 }
 
-
-
 template < typename T >
-frustum< T >::~frustum()
-{}
-
-
-
-template< typename T >
-frustum< T >& frustum< T >::operator=( const frustum& source_ )
+Frustum<T>::Frustum( const matrix< 4, 4, T >& projection )
 {
-    memcpy( array, source_.array, 6 * sizeof( T ) );
-    return *this;
+    _array[4] = projection( 2, 3 ) / (projection( 2, 2 ) - 1.0);
+    _array[5] = projection( 2, 3 ) / (projection( 2, 2 ) + 1.0);
+    _array[2] = nearPlane() * ( projection( 1, 2 ) - 1.0 ) / projection( 1, 1 );
+    _array[3] = nearPlane() * ( projection( 1, 2 ) + 1.0 ) / projection( 1, 1 );
+    _array[0] = nearPlane() * ( projection( 0, 2 ) - 1.0 ) / projection( 0, 0 );
+    _array[1] = nearPlane() * ( projection( 0, 2 ) + 1.0 ) / projection( 0, 0 );
```

Comment:
l.181?

the b,t,l,r planes are specified wrt the front plane. See also adjustNear().